Serizalize model only once to reduce backend preparation overhead

IceTDrinker opened the pull request:
**Description**: Serizalize model only once to reduce backend preparation overhead

**Motivation and Context**
- The serialization can be very heavy for large models, this avoids doing it twice

IceTDrinker commented:
Hello @thiagocrepaldi , sorry if this is not the way to request a review, but I could not find who the code owners for the files in this PR are.

I saw no guidelines as to how one should request a review (being a microsoft outsider does not help 😅 ) and you seemed to be the person indicated for most python code related to training (though this is not training, it's a small python patch).

CONTRIBUTING.md had no indication as to how review should be requested, would you kindly indicate how/who I should request a review from ?

edgchen1 commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux CPU x64 NoContribops CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux Nuphar CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline

edgchen1 commented:
/azp run MacOS NoContribops CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows WebAssembly CI Pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

edgchen1 commented:
@xadupre mind taking a look?

IceTDrinker commented:
Sorry for the inconvenience, I see some pipelines failing, but it's not pulling onnx 1.10.x but rather 1.9.x which does not seem to have the functionality, as the submodule just indicates the repository I thought the latest onnx package was being used.

Probably a requirements.txt that needs updating, guidance welcome.

IceTDrinker commented:
Among required pipelines :

Submodule failed :
- Linux CPU x64 NoContribops CI Pipeline
- Linux GPU CI Pipeline
- Linux OpenVINO CI Pipeline
- Windows CPU CI Pipeline
- Windows WebAssembly CI Pipeline

Genuine failure due to bad onnx package :
- Linux CPU CI Pipeline
- orttraining-linux-ci-pipeline

Offending commands during dep installation from e.g. (https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=456755&view=logs&j=7536d2cd-87d4-54fe-4891-bfbbf2741d83&t=32a004b5-6abc-5719-dbf1-16350e4ba258):

<details><summary>Logs</summary>
<p>

```console
+ python3 -m pip uninstall -y ort-nightly-gpu ort-nightly onnxruntime onnxruntime-gpu -qq
+ cp /mnt/vss/_work/2/s/tools/ci_build/github/linux/docker/scripts/manylinux/requirements.txt /mnt/vss/_work/2/b/requirements.txt
+ sed -i 's/git+http:\/\/github\.com\/onnx\/onnx.*/onnx==1.9.0/' /mnt/vss/_work/2/b/requirements.txt
+ python3 -m pip install -r /mnt/vss/_work/2/b/requirements.txt
```

</p>
</details>


How should the onnx python dependency be updated  ?

IceTDrinker commented:
Just updated the PR and modified workflows to install onnx 1.10.1 as the onnx submodule is already checked out to this version.

IceTDrinker commented:
Gentle ping @edgchen1

tianleiwu commented on tools/ci_build/github/azure-pipelines/linux-ci-pipeline.yml:
```diff
@@ -54,7 +54,7 @@ jobs:
          cp $(Build.SourcesDirectory)/tools/ci_build/github/linux/docker/scripts/manylinux/requirements.txt $(Build.BinariesDirectory)/requirements.txt
          #Test ORT with the latest ONNX release.
-         sed -i 's/git+http:\/\/github\.com\/onnx\/onnx.*/onnx==1.9.0/' $(Build.BinariesDirectory)/requirements.txt
+         sed -i 's/git+http:\/\/github\.com\/onnx\/onnx.*/onnx==1.10.1/' $(Build.BinariesDirectory)/requirements.txt
```
Does this PR has dependency on latest onnx version, and not supported by 1.9.0?

I did not see 1.10.1 supported in the following table:
https://github.com/microsoft/onnxruntime/blob/master/docs/Versioning.md#version-matrix

IceTDrinker replied:
Yes, my change does have a dependency on latest onnx 1.10.1 and the onnx submodule was set to 1.10.1 in the source tree when I did the last change hence the 1.10.1 version change here.

IceTDrinker replied:
done @tianleiwu thanks for the review

tianleiwu commented:
/azp run Linux CPU x64 NoContribops CI Pipeline, Linux GPU CI Pipeline

tianleiwu commented:
@IceTDrinker, some build pipeline failed. Could you integrate the latest master to your branch? Thanks.

IceTDrinker commented:
> @IceTDrinker, some build pipeline failed. Could you integrate the latest master to your branch? Thanks.

Actually the failure is due to the new version check, I just realized we can get an infinite recursion and it shows in tests. I'll see if there is another way to write this.

IceTDrinker commented:
@tianleiwu I saw you wrote a comment about how to check for the package version but it could fail if the `packaging` is not installed and I don't know if it is a dependency for onnxruntime, what do you want me to do in that case ?

tianleiwu commented:
> @tianleiwu I saw you wrote a comment about how to check for the package version but it could fail if the `packaging` is not installed and I don't know if it is a dependency for onnxruntime, what do you want me to do in that case ?

That's right, `packaging` is not a dependency of onnxruntime.

I added the comment since onnx has weekly build like "onnx-weekly 1.10.1.dev20211011". That will get exception like `ValueError: invalid literal for int() with base 10: 'dev20211011'`. I think you can use `version.version.split(".")[:3]` to avoid it.

IceTDrinker commented:
> > @tianleiwu I saw you wrote a comment about how to check for the package version but it could fail if the `packaging` is not installed and I don't know if it is a dependency for onnxruntime, what do you want me to do in that case ?
> 
> That's right, `packaging` is not a dependency of onnxruntime.
> 
> I added the comment since onnx has weekly build like "onnx-weekly 1.10.1.dev20211011". That will get exception like `ValueError: invalid literal for int() with base 10: 'dev20211011'`. I think you can use `version.version.split(".")[:3]` to avoid it.

Yes, I let the pipelines finish to check the code is finally sound and will add the [:3] afterwards
